README.md
[![published](https://static.production.devnetcloud.com/codeexchange/assets/images/devnet-published.svg)](https://developer.cisco.com/codeexchange/github/repo/CiscoDevNet/homebridge-c9800)
[![npm](https://badgen.net/npm/v/homebridge-c9800)](https://www.npmjs.com/package/homebridge-c9800)[![npm](https://badgen.net/npm/dw/homebridge-c9800)](https://www.npmjs.com/package/homebridge-c9800)[![npm](https://badgen.net/npm/dt/homebridge-c9800)](https://www.npmjs.com/package/homebridge-c9800)

# A [Cisco](https://www.cisco.com) Wireless LAN Controller plugin for [Homebridge](https://github.com/nfarina/homebridge).  

---

# Purpose

The purpose of this code is to create a Homebridge plugin to allow a wireless administrator to enable or disable a WLAN through a Light Switch control in Apple’s homekit or through Siri.

![alt text](example.png "Dashboard Example Screenshot")

![alt text](example2.png "WLC Example Screenshot")

# Intended Audience

This code was originally designed for my personal use at home, to allow me to toggle my guest wireless network on and off as needed.  This example could be helpful for those who are:

- Wishing to learn how to use RESTCONF
- Considering methods of automating configuration changes on the C9800 WLC
- Interested in integrating Cisco gear into their home automation

# How This Code Works

This NodeJS code will accomplish the following tasks:
1. Creates a parent container accessory in Homebridge, to be shared with Apple homekit - one per WLC
2. Creates a set of child Light Switch accessories in Homebridge, to be shared with Apple homekit - one per WLAN
3. Exposes two functions to Homebridge – one to interrogate the WLAN status, the other to set the status
4. Provides any error back to Homebridge for logging


# Installation Steps
Run these commands:

    sudo npm install -g homebridge
    sudo npm install -g homebridge-c9800


NOTE: If you install homebridge like this:

    sudo npm install -g --unsafe-perm homebridge

Then all subsequent plugin installations must be like this:

    sudo npm install -g --unsafe-perm homebridge-c9800

# Configuration
Example platform config (needs to be added to the homebridge config.json):
```
  "platforms": [
    {
      "name": "Platform C9800",
      "timeout": 10000,
      "refreshInterval": 15,
      "debug": false;
      "wlcs": [
        {
          "displayName": "Building 1 WLC",
          "ipAddress": "192.168.1.110",
          "username": "admin",
          "password": "password"
        },
        {
          "displayName": "Building 2 WLC",
          "ipAddress": "192.168.1.115",
          "username": "admin",
          "password": "password"
        }            
      ],
      "platform": "Cisco 9800"
    }
  ]
```

### Config Explanation:

Field           						| Description
----------------------------|------------
**platform**   							| (required) Must always be "Cisco 9800".
**name**										| (required) The internal name you want to use for the platform within Homebridge.
**timeout**									| (optional) The timeout duration in ms for the web API calls.
**refreshInterval**					| (optional) The refresh interval in minutes for polling the WiFi status.
**debug**										| (optional) Enables additional logging.
**displayName**							| (required) The name you want to show for the WLC in homekit.
**ipAddress**								| (required) The IP address of the WLC (should be static, not DHCP).
**username**								| (required) The username used to access the WLC.
**password**								| (required) The password used to access the WLC.

To make your WLC work with the plugin:

1. Connect your WLC to your network.
2. Configure your WLC.
3. Configure "aaa authorization exec default local" or equivalent on your WLC.
4. Enable restconf on the WLC.
5. Write down the IP address of the WLC.
6. Create your config file according to the above example (or using the Homebridge UI).

# FAQ
1. What is the purpose of each file?
   - [index.js](index.js) - Contains the NodeJS Homebridge plugin registration
   - [c9800platform.js](c9800platform.js) - Contains the NodeJS Homebridge plugin code for the creating the WLC containers and instantiating the WLAN LightSwitches
   - [c9800platformAccessory.js](c9800platformAccessory.js) - Contains the NodeJS Homebridge plugin code for the WLAN LightSwitches
   - [config.schema.json](config.schema.json) - Contains the schema to allow configuration of the plugin through the Homebridge UI
2. Does this code use NETCONF, RESTCONF, or both?
   - This code leverages RESTCONF APIs and YANG data models only. NETCONF is not used.
3. How do I enable RESTCONF on my Catalyst 9800 WLC?
   - From a command prompt, type:
      ```console
      WLC(config)# restconf
      ```
   - More information can be found [here](https://developer.cisco.com/docs/ios-xe/#!enabling-restconf-on-ios-xe/authentication)
4. How do I properly modify the Homebridge config.json file with the appropriate information?
   - See the config explanation section above.
5. Will this work with my model of Wireless LAN Controller?
   - This was tested to work with the Cisco Catalyst 9800CL. It should work with any other model of Cisco Catalyst 9800.
   - This has also been tested on the Cisco Embedded Wireless LAN Controller on the C9100-series access points.


# More information
Check out https://github.com/nfarina/homebridge for more information about Homebridge.



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## 0.1.4 (2021-12-31)

#### Changes

- Fixed a polling bug for IOS XE 17.7.1.
- Updated dependencies.

## 0.1.3 (2021-11-09)

#### Changes

- Fixed a polling bug.

## 0.1.2 (2021-11-08)

#### Changes

- Changed to interval polling.

## 0.1.1 (2021-10-28)

#### Changes

- Dynamically load Model, SerialNumber & FirmwareVersion from the WLC.
- Removed config entries for model and serial.

## 0.1.0 (2021-10-26)

#### Changes

- Converted to a dynamic platform, to allow multiple WLCs to be used, with multiple SSIDs on each.

## 0.0.5 (2021-10-14)

#### Changes

- Readability.

## 0.0.4 (2021-10-14)

#### Changes

- Documentation and readability.
- Updated dependencies.

## 0.0.3 (2021-10-14)

#### Changes

- Added clarification on WLC's SSID status response in debug.
- Copyright changed to Cisco Systems.

## 0.0.2 (2021-10-08)

#### Changes

- Migration to my Github repository.

## 0.0.1 (2021-10-05)

#### Changes

- Initial commit.


c9800platformAccessory.js
'use strict'
const wlan_cfg_entries = "/restconf/data/Cisco-IOS-XE-wireless-wlan-cfg:wlan-cfg-data/wlan-cfg-entries"
const wlan_cfg_path = wlan_cfg_entries + "/wlan-cfg-entry="

/**
 * Platform Accessory
 * An instance of this class is created for each accessory your platform registers
 * Each accessory may expose multiple services of different service types.
 */
class C9800PlatformAccessory {
//private service: Service;

  constructor(platform,accessory){
  	this.platform = platform
  	this.accessory = accessory
  	this.services = []
  	  	
    for (let i = 0; i < this.accessory.context.device.wlanCount; i++) {
	    const iString = i.toString()
	    
	    const switchService = this.accessory.getService(iString) ?? this.accessory.addService(this.platform.Service.Switch, this.accessory.context.device.ssidList[i], iString)
      // To avoid "Cannot add a Service with the same UUID another Service without also defining a unique 'subtype' property." error,
      // when creating multiple services of the same type, you need to use the following syntax to specify a name and subtype id:
      // this.accessory.getService('NAME') ?? this.accessory.addService(this.platform.Service.Switch, 'NAME', 'USER_DEFINED_SUBTYPE');

      // register handlers for the On/Off Characteristic
      switchService.getCharacteristic(this.platform.Characteristic.On)
        .on('set', this.setPowerState.bind(this, i))   // SET - bind to the `setPowerState` method below
    }
 
    // set accessory information
    this.accessory.getService(this.platform.Service.AccessoryInformation)
      .setCharacteristic(this.platform.Characteristic.Manufacturer, 'Cisco')
      .setCharacteristic(this.platform.Characteristic.Model, this.accessory.context.device.model)
      .setCharacteristic(this.platform.Characteristic.SerialNumber, this.accessory.context.device.serial)
      .setCharacteristic(this.platform.Characteristic.FirmwareRevision, this.accessory.context.device.firmware)
		this.timer = setTimeout(this.poll.bind(this), this.platform.refreshInterval)
		this.poll()

	}

	async setPowerState(index, state, callback) {
		const context = this.accessory.context
		try {
			this.platform.log('Setting WLAN state for %s = %s', this.accessory.context.device.ssidList[index], state)
			const response = await this.platform.session({
				method: 'patch',
				url: 'https://'+context.device.ipAddress+wlan_cfg_path+this.accessory.context.device.wlanList[index]+'/apf-vap-id-data/wlan-status',
				data: {"Cisco-IOS-XE-wireless-wlan-cfg:wlan-status": state},
				auth: {
					username: context.device.username,
					password: context.device.password
				},
				headers: {
					'Accept': 'application/yang-data+json',
					'Content-Type': 'application/yang-data+json'
				},
				timeout: this.platform.timeout
			}).catch(err => {
					this.platform.log.error('Error setting WLAN state %s',err)
					this.platform.log.warn(err.response.data.errors)
			})
			callback (null)
		}catch(err) {
			this.platform.log.error('Error setting WLAN state %s', err)
			callback (err)
		}
	}

	
	async poll() {
		if(this.timer) clearTimeout(this.timer)
		this.timer = null
		for (let i = 0; i < this.accessory.context.device.wlanCount; i++) {
			try {
				const response = await this.platform.session({
					method: 'get',
					url: 'https://' + this.accessory.context.device.ipAddress+wlan_cfg_path + this.accessory.context.device.wlanList[i]+'/apf-vap-id-data/wlan-status',
					data: {},
					auth: {
						username: this.accessory.context.device.username,
						password: this.accessory.context.device.password
					},
					headers: {
						'Accept': 'application/yang-data+json',
						'Content-Type': 'application/yang-data+json'
					},
					timeout: this.platform.timeout
				}).catch(err => {
						this.platform.log.error('Error getting WLAN state %s',err)
				})
				const powerState = response["data"]["Cisco-IOS-XE-wireless-wlan-cfg:wlan-status"];
				this.platform.log('Get WLAN state for %s = %s',  this.accessory.context.device.ssidList[i], (powerState === true))
				this.accessory.services[i+1].getCharacteristic(Characteristic.On).updateValue(powerState === true)
			}catch(err) {
				this.platform.log.error('Error getting WLAN state %s', err)
			}
		}
		this.timer = setTimeout(this.poll.bind(this), this.platform.refreshInterval)	
	}
}
module.exports=C9800PlatformAccessory;


c9800platform.js
'use strict'

//
//	We are going to use axios with SSL for our RESTconf calls to the WLC,
//	so we'll need these node-modules

const https = require('https')
const axios = require('axios')
axios.defaults.withCredentials = true
const packageJson=require('./package')
const C9800PlatformAccessory=require('./c9800platformAccessory')

//
//	Setting up a few constant variables for reference

const wlan_cfg_entries = "/restconf/data/Cisco-IOS-XE-wireless-wlan-cfg:wlan-cfg-data/wlan-cfg-entries"
const wlan_cfg_path = wlan_cfg_entries + "/wlan-cfg-entry="
const wlc_device_inventory = "/restconf/data/Cisco-IOS-XE-device-hardware-oper:device-hardware-data/device-hardware/device-inventory"
const wlc_software_version = "/restconf/data/Cisco-IOS-XE-device-hardware-oper:device-hardware-data/device-hardware/device-system-data/software-version"
const timeout = 10000
const interval = 15 // Minutes
const	PluginName = packageJson.name
//const	PlatformName = PLATFORM_NAME

class C9800Platform {

	//
	//	Setup our instance variables based on the homebridge config.json entries

	constructor(log, config, api) {

		this.log = log
		this.config = config
		this.name = config["name"]
		this.timeout = config["timeout"] === undefined ? timeout : config["timeout"]
		this.refreshInterval = config["refreshInterval"] === undefined ? (interval * 60000) : (config["refreshInterval"] * 60000)
		this.debug = config["debug"] === undefined ? false : config["debug"]
		this.accessories = []
		this.session = axios.create({
			timeout: this.timeout,
			httpsAgent: new https.Agent({  
				rejectUnauthorized: false
				}),
		})

		if(api){
			this.api=api
			this.Service = this.api.hap.Service
  		this.Characteristic = this.api.hap.Characteristic
			this.api.on("didFinishLaunching", function(){
				//Get devices
				this.getC9800Devices()
			}.bind(this))
		}
	}
	
	identify(){
		this.log('Identify the WLC')
	}
	
	configureAccessory(accessory) {
    if (this.config.wlcs.find( (wlc) => wlc.ipAddress === accessory.context.device.ipAddress ))  {
			if (!this.accessories.find( (anAccessory) => anAccessory.context.device.ipAddress === accessory.context.device.ipAddress ))  {
				this.log.info('Restoring accessory from cache:', accessory.displayName)

				// create the accessory handler
				// this is imported from `c9000platformAccessory.js`
				new C9800PlatformAccessory(this, accessory)

				// add the restored accessory to the accessories cache so we can track if it has already been registered
				this.accessories.push(accessory)
			}
    }
	}
	
	getC9800Devices(){

		//
    // Check for a blank config and return without registering accessories

		if(!this.config){
			this.log.warn('Ignoring C9800 Platform setup because it is not configured')
			return
		}
		
    // loop over the discovered devices and register each one if it has not already been registered
		for (const device of this.config.wlcs){
			this.log.debug('Getting WLAN list')
			var wlcDevice = {
				"displayName": device.displayName,
				"ipAddress": device.ipAddress,
				"username": device.username,
				"password": device.password,
				"model": "",
				"serial": "",
				"firmware": "0.1",
				"wlanCount": 0,
				"wlanList": [],
				"ssidList": []
			}
			const webConfig = {
				data: {},
				auth: {
					username: device.username,
					password: device.password
				},
				headers: {
					'Accept': 'application/yang-data+json',
					'Content-Type': 'application/yang-data+json'
				},
			};
			this.session.get('https://'+device.ipAddress+wlc_device_inventory, webConfig)
				.then(response => {
					var wlcModel = response.data["Cisco-IOS-XE-device-hardware-oper:device-inventory"][0]["part-number"]
					var wlcSerial = response.data["Cisco-IOS-XE-device-hardware-oper:device-inventory"][0]["serial-number"]
					this.log('WLC model: %s, serial: %s',wlcModel,wlcSerial)
			
					wlcDevice.model = wlcModel === "" ? "C9800" : wlcModel
					wlcDevice.serial = wlcSerial === "" ? "DummySerial" : wlcSerial

					this.session.get('https://'+device.ipAddress+wlc_software_version, webConfig)
						.then(resp2 => {
							var wlcFirmware = resp2.data["Cisco-IOS-XE-device-hardware-oper:software-version"]
							wlcFirmware = wlcFirmware.substring((wlcFirmware.search("Version ")+8))
							wlcFirmware = wlcFirmware.substring(0,wlcFirmware.search(","))
							this.log('WLC firmware: %s',wlcFirmware)
			
							wlcDevice.firmware = wlcFirmware === "" ? "0.1" : wlcFirmware

							this.session.get('https://'+device.ipAddress+wlan_cfg_entries, webConfig)
								.then(resp3 => {
									wlcDevice.wlanCount = Object.keys(resp3.data["Cisco-IOS-XE-wireless-wlan-cfg:wlan-cfg-entries"]["wlan-cfg-entry"]).length
									this.log("WLC response: " + wlcDevice.wlanCount + " WLAN(s).")
									const wlans = resp3.data["Cisco-IOS-XE-wireless-wlan-cfg:wlan-cfg-entries"]["wlan-cfg-entry"]
									for (var i = 0; i < wlcDevice.wlanCount; i++) {
										if (this.debug) {
											this.log(i, wlans[i]["profile-name"] + " = " + wlans[i]["apf-vap-id-data"]["ssid"])
										}
										var wlan = wlans[i]["profile-name"]
										var ssid = wlans[i]["apf-vap-id-data"]["ssid"]
										wlcDevice.wlanList[i]=wlan
										wlcDevice.ssidList[i]=ssid
										this.log(`SSID ${ssid}`, i)
									}

									// generate a unique id for the accessory this should be generated from
									// something globally unique, but constant, for example, the device serial
									// number or MAC address
									const uuid = this.api.hap.uuid.generate(device.ipAddress);

									// check that the device has not already been registered by checking the
									// cached devices we stored in the `configureAccessory` method above
									if (!this.accessories.find(accessory => accessory.UUID === uuid)) {
										this.log.info('Registering new accessory:', device.displayName, uuid)

										// create a new accessory
										const accessory = new this.api.platformAccessory(device.displayName, uuid)

										// store a copy of the device object in the `accessory.context`
										// the `context` property can be used to store any data about the accessory you may need
										accessory.context.device = wlcDevice
						
										for (let i=0; i<accessory.context.device.wlanCount; i++){
											if (this.debug) {this.log.info(i, accessory.context.device.wlanList[i] + " = " + accessory.context.device.ssidList[i])}
										}

										// create the accessory handler
										// this is imported from `platformAccessory.js`
										new C9800PlatformAccessory(this, accessory)

										// link the accessory to your platform
										this.api.registerPlatformAccessories(PluginName, PlatformName, [accessory])

										// push into accessory cache
										this.accessories.push(accessory)
										this.log.info('New accessory:', accessory.displayName)

										// it is possible to remove platform accessories at any time using `api.unregisterPlatformAccessories`, eg.:
										// this.api.unregisterPlatformAccessories(PluginName, PlatformName, [accessory]);
									}
								})
								.catch(err => {
									this.log.error(err.stack)
								})
						})
						.catch(err => {
							this.log.error(err.stack)
						})
				})
				.catch(err => {
					this.log.error(err.stack)
				})
		}
	}
}

module.exports=C9800Platform;